gems/sorbet-runtime/lib/types/private/methods/_methods.rb

```diff
@@ -346,11 +350,31 @@ def self.run_sig(hook_mod, method_name, original_method, declaration_block)
   end
 
   def self.run_builder(declaration_block)
-    builder = DeclBuilder.new(declaration_block.mod, declaration_block.raw)
-    builder
-      .instance_exec(&declaration_block.blk)
-      .finalize!
-      .decl
+    return if has_live_decl_builder?(declaration_block)
+
+    builder = register_live_decl_builder(declaration_block)
+    begin
+      builder.run!(&declaration_block.blk)
+    ensure
+      unregister_live_decl_builder(declaration_block)
+    end
+
+    builder.decl
+  end
+
+  @live_decl_builders = {}
+  def self.has_live_decl_builder?(declaration_block)
+    @live_decl_builders[declaration_block.mod.hash]
+  end
+
+  def self.register_live_decl_builder(declaration_block)
+    @live_decl_builders[declaration_block.mod.hash] = true
+
+    declaration_block.decl_builder
+  end
+
+  def self.unregister_live_decl_builder(declaration_block)
+    @live_decl_builders[declaration_block.mod.hash] = false
```

Comment:
@froydnj This is our alternative solution to handle the edge case described in the PR documentation and the test case. You can look at my earlier response [here](https://github.com/sorbet/sorbet/pull/5329#discussion_r822086310) to  gain more context.

Instead of handling the edge case like the current implementation on `master`, we decided to keep track of the `DeclBuilder` that is being executed. We store them in a hash using [Object#hash](https://ruby-doc.org/core-3.1.1/Object.html#hash-method) as the key and ensure the same builder isn't already being processed for any given execution.

The early return inside `run_builder` prevents the edge case from happening entirely rather than dealing with it after the fact unlike the previous implementation.

Reply:
Thanks for looking into this @KaanOzkan!

Two issues with this implementation:

1. `@live_decl_builders` grows essentially without bound.  I guess it's bounded by the number of signatures in the program, but that can still be pretty substantial.  I think a better implementation would remove the entry in the hash table in the `ensure` block.
2. The documentation for `Object.hash` says that `a.eql?(b)` implies `a.hash == b.hash`, but the reverse is not true: `a.hash == b.hash` doesn't imply anything about the equality of `a` and `b`.  So I would be worried that we could have `DeclarationBlock`s for two different modules, the modules have the same hash, and we would be blocked from running `sig`s for both modules at the same time.  I guess the table is indexed by `mod.hash` instead of by `mod` in the interest of not retaining references to the modules...or the declaration block?  Fortunately, the solution to the first problem would fix the second problem.

Reply:
(I still think it ought to be possible to write things such that we don't need re-entrancy hacks like this, since the previous implementation of `sig` was able to get by without it...)